language/move-lang/src/hlir/translate.rs

```diff
@@ -0,0 +1,1227 @@
+        TE::Value(v) => HE::Value(v),
+        TE::Move { from_user, var } => HE::Move {
+            from_user,
+            var: context.remapped_local(var),
+        },
+        TE::Copy { from_user, var } => HE::Copy {
+            from_user,
+            var: context.remapped_local(var),
+        },
+        TE::Use(_) => panic!("ICE unexpanded use"),
+        TE::ModuleCall(call) => {
+            let T::ModuleCall {
+                module,
+                name,
+                type_arguments,
+                arguments,
+                parameter_types,
+                acquires,
+            } = *call;
+            let a_m_f = (
+                &module.0.value.address,
+                module.0.value.name.value(),
+                name.value(),
+            );
+            if let (&Address::LIBRA_CORE, "Transaction", "assert") = a_m_f {
+                let tbool = N::SingleType_::bool(eloc);
+                let tu64 = N::SingleType_::u64(eloc);
+                let tunit = sp(eloc, N::Type_::Unit);
+                let vcond = Var(sp(eloc, new_name("tmp%")));
+                let vcode = Var(sp(eloc, new_name("tmp%")));
+
+                let mut stmts = VecDeque::new();
+
+                let bvar = |v, st| sp(eloc, T::Bind_::Var(v, Some(st)));
+                let bind_list = sp(
+                    eloc,
+                    vec![
+                        bvar(vcond.clone(), tbool.clone()),
+                        bvar(vcode.clone(), tu64.clone()),
+                    ],
+                );
+                let tys = vec![Some(tbool.clone()), Some(tu64.clone())];
+                let bind = sp(eloc, T::SequenceItem_::Bind(bind_list, tys, arguments));
+                stmts.push_back(bind);
+
+                let mvar = |var, st| {
+                    let from_user = false;
+                    let mv = TE::Move { from_user, var };
+                    T::exp(N::Type_::single(st), sp(eloc, mv))
+                };
+                let econd = mvar(vcond, tu64);
+                let ecode = mvar(vcode, tbool);
+                let eabort = T::exp(tunit.clone(), sp(eloc, TE::Abort(Box::new(ecode))));
+                let eunit = T::exp(tunit.clone(), sp(eloc, TE::Unit));
+                let inlined_ = TE::IfElse(Box::new(econd), Box::new(eunit), Box::new(eabort));
+                let inlined = T::exp(tunit.clone(), sp(eloc, inlined_));
+                stmts.push_back(sp(eloc, T::SequenceItem_::Seq(Box::new(inlined))));
+
+                let block = T::exp(tunit, sp(eloc, TE::Block(stmts)));
+                return maybe_exp_(context, result, block);
+            }
+            let mut expected_types = single_types(context, parameter_types);
+            let expected_type = match expected_types.len() {
+                0 => sp(eloc, H::Type_::Unit),
+                1 => sp(eloc, H::Type_::Single(expected_types.pop().unwrap())),
+                _ => sp(eloc, H::Type_::Multiple(expected_types)),
+            };
```

Comment:
Do you want to call `expected_types` here (/the line above)? 
If not, might be worthwhile adding a helping function for this so that you can combine this with the code in `expected_types`